bot/exts/info/subscribe.py

```diff
@@ -0,0 +1,149 @@
+import arrow
+import discord
+from discord.ext import commands
+from discord.interactions import Interaction
+
+from bot import constants
+from bot.bot import Bot
+from bot.decorators import in_whitelist
+from bot.log import get_logger
+from bot.utils import checks, members, scheduling
+
+# Tuple of tuples, where each inner tuple is a role id and a month number.
+# The month number signifies what month the role should be assignable,
+# use None for the month number if it should always be active.
+ASSIGNABLE_ROLES = (
+    (constants.Roles.announcements, None),
+    (constants.Roles.pyweek_announcements, None),
+    (constants.Roles.advent_of_code, 12),
+    (constants.Roles.lovefest, 2),
+)
+ITEMS_PER_ROW = 3
+
+log = get_logger(__name__)
+
+
+class RoleButtonView(discord.ui.View):
+    """A list of SingleRoleButtons to show to the member."""
+
+    def __init__(self, member: discord.Member):
+        super().__init__()
+        self.interaction_owner = member
+
+    async def interaction_check(self, interaction: Interaction) -> bool:
+        """Ensure that the user clicking the button is the member who invoked the command."""
+        if interaction.user != self.interaction_owner:
+            await interaction.response.send_message(
+                ":x: This is not your command to react to!",
+                ephemeral=True
+            )
+            return False
+        return True
+
+
+class SingleRoleButton(discord.ui.Button):
+    """A button that adds or removes a role from the member depending on it's current state."""
+
+    ADD_STYLE = discord.ButtonStyle.success
+    REMOVE_STYLE = discord.ButtonStyle.secondary
+    LABEL_FORMAT = "{action} role {role_name}"
+    CUSTOM_ID_FORMAT = "subscribe-{role_id}"
+
+    def __init__(self, role: discord.Role, assigned: bool, row: int):
+        super().__init__(
+            style=self.REMOVE_STYLE if assigned else self.ADD_STYLE,
+            label=self.LABEL_FORMAT.format(action="Remove" if assigned else "Add", role_name=role.name),
+            custom_id=self.CUSTOM_ID_FORMAT.format(role_id=role.id),
+            row=row,
+        )
+        self.role = role
+        self.assigned = assigned
+
+    async def callback(self, interaction: Interaction) -> None:
+        """Update the member's role and change button text to reflect current text."""
+        if isinstance(interaction.user, discord.User):
+            await interaction.message.delete()
+            self.view.stop()
+            return
```

Comment:
Maybe we should add a trace or debug call when this happens?

Reply:
https://github.com/python-discord/bot/pull/1868/commits/009b2bb8dc1fc3442c076ef4ecf66d6dc77eae4e